File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/Protocols/FLTRequestHandlerProtocol.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "FLTRequestHandlerProtocol.h"
#import <Foundation/Foundation.h>
#import "FIAPRequestHandler.h"

@interface DefaultRequestHandler ()
/// The wrapped FIAPRequestHandler
@property(nonatomic, strong) FIAPRequestHandler *handler;
@end

@implementation DefaultRequestHandler

- (void)startProductRequestWithCompletionHandler:(nonnull ProductRequestCompletion)completion {
  [self.handler startProductRequestWithCompletionHandler:completion];
}

- (nonnull instancetype)initWithRequestHandler:(nonnull FIAPRequestHandler *)handler {
  self = [super init];
  if (self) {
    _handler = handler;
  }
  return self;
}
@end

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit/Protocols/FLTRequestHandlerProtocol.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation
import StoreKit

/// A protocol that wraps SKRequest.
@objc public protocol FLTRequestHandlerProtocol: NSObjectProtocol {
  /// Wrapper for SKRequest's start
  /// https://developer.apple.com/documentation/storekit/skrequest/1385534-start
  func startProductRequest(completionHandler: @escaping (SKProductsResponse?, Error?) -> Void)
}

/// The default request handler that wraps FIAPRequestHandler
public class DefaultRequestHandler: NSObject, FLTRequestHandlerProtocol {
  /// The wrapped FIAPRequestHandler
  private let handler: FIAPRequestHandler

  /// Initialize this wrapper with an instance of FIAPRequestHandler
  public init(requestHandler: FIAPRequestHandler) {
    self.handler = requestHandler
  }

  public func startProductRequest(
    completionHandler: @escaping (SKProductsResponse?, Error?) -> Void
  ) {
    handler.startProductRequest(completionHandler: completionHandler)
  }
}
